streams/src/test/java/org/apache/kafka/streams/kstream/internals/KTableTransformValuesTest.java

```diff
@@ -0,0 +1,553 @@
+                                .withKeySerde(Serdes.String())
+                                .withValueSerde(Serdes.Integer()))
+                .groupBy(toForceSendingOfOldValues(), Serialized.with(Serdes.String(), Serdes.Integer()))
+                .reduce(MockReducer.INTEGER_ADDER, MockReducer.INTEGER_SUBTRACTOR)
+                .mapValues(mapBackToStrings())
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aa", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aaa", 0L));
+
+        assertThat(output(), hasItems("A:1", "A:0", "A:2", "A:0", "A:3"));
+
+        final KeyValueStore<String, Integer> keyValueStore = driver.getKeyValueStore(QUERYABLE_NAME);
+        assertThat(keyValueStore.get("A"), is(3));
+    }
+
+    @Test
+    public void shouldCalculateCorrectOldValuesIfNotStatefulEvenIfNotMaterialized() {
+
+        builder.addStateStore(storeBuilder(STORE_NAME))
+                .table(INPUT_TOPIC, CONSUMED)
+                .transformValues(new StatelessTransformerSupplier())
+                .groupBy(toForceSendingOfOldValues(), Serialized.with(Serdes.String(), Serdes.Integer()))
+                .reduce(MockReducer.INTEGER_ADDER, MockReducer.INTEGER_SUBTRACTOR)
+                .mapValues(mapBackToStrings())
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aa", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aaa", 0L));
+
+        assertThat(output(), hasItems("A:1", "A:0", "A:2", "A:0", "A:3"));
+    }
+
+    @Test
+    public void willUnfortunatelyCalculateIncorrectOldValuesIfStatefulAndNotMaterialized() {
+        builder.addStateStore(storeBuilder(STORE_NAME))
+                .table(INPUT_TOPIC, CONSUMED)
+                .transformValues(new StatefulTransformerSupplier())
+                .groupBy(toForceSendingOfOldValues(), Serialized.with(Serdes.String(), Serdes.Integer()))
+                .reduce(MockReducer.INTEGER_ADDER, MockReducer.INTEGER_SUBTRACTOR)
+                .mapValues(mapBackToStrings())
+                .toStream()
+                .process(capture);
+
+        driver = new TopologyTestDriver(builder.build(), props());
+
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "a", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aa", 0L));
+        driver.pipeInput(recordFactory.create(INPUT_TOPIC, "A", "aaa", 0L));
+
+        assertThat(output(), not(hasItems("A:1", "A:0", "A:2", "A:0", "A:3")));
+        // Output more likely to be "A:1", "A:-2", "A:0", "A:-5", "A:-1"
```

Comment:
Is this test useful?

Reply:
well, indeed. I did call it out. It does highlight the issue and, if it starts to fail, there is a doc update needed. Happy to remove though. (I agree its a bit weird.)

Reply:
The idea with the doc change is good -- maybe put a comment about this -- otherwise, when it fails, we might not remember the purpose of this test.